fix(core): incorrectly inserting elements inside <template> element

crisbeto opened the pull request:
Currently whenever we insert elements, we do it directly on the node using `appendChild` or `insertBefore`, however this doesn't work with `<template>` elements where the `template.content` has to be used.

These changes add a few checks to call `appendChild` and `insertBefore` on the `content` of the template.

Fixes #15557.

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -613,6 +613,8 @@ export function nativeInsertBefore(
   if (isProceduralRenderer(renderer)) {
     renderer.insertBefore(parent, child, beforeNode, isMove);
+  } else if (isTemplateNode(parent)) {
+    parent.content.appendChild(child);
```
Since `parent.content` is a document fragment, should we call `insertBefore` too?

```suggestion
    parent.content.insertBefore(child, beforeNode, isMove);
```

Or there is a risk that `beforeNode` might be outside of the `<template>` element?

It just seems a bit strange that we have different logic here for the template (always append) vs other elements (insert after).

crisbeto replied:
ops, good catch. This was supposed to be `insertBefore`, but I must have been copying things around too quickly 😅

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -623,6 +625,8 @@ function nativeAppendChild(renderer: Renderer3, parent: RElement, child: RNode):
   ngDevMode && assertDefined(parent, 'parent node must be defined');
   if (isProceduralRenderer(renderer)) {
     renderer.appendChild(parent, child);
+  } else if (isTemplateNode(parent)) {
+    parent.content.appendChild(child);
   } else {
     parent.appendChild(child);
   }
```
Nit: this should minify a bit better:
```
  } else {
    const parentEl = isTemplateNode(parent) ? parent.content : parent;
    parentEl.appendChild(child);
  }
```

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -647,6 +651,11 @@ function nativeRemoveChild(
   }
 }
 
+/** Gets if an element is a `<template>` node. */
```
```suggestion
/** Checks if an element is a `<template>` node. */
```

crisbeto commented:
The feedback has been addressed.

lpl14 commented:
There are duplicate entries in multiple _bundle.golden_symbols.json_

`  {
    "name": "isTemplateNode"
  },`

Is this intentional?

crisbeto commented:
I believe the duplicate entries are due to the fact that I have functions called `isTemplateNode` both in `core` and `platform-browser`.

lpl14 commented:
@crisbeto  But the duplicates are even within the same file?
E.g. in _packages/core/test/bundling/forms_reactive/bundle.golden_symbols.json_ from line 1238 to 1243.

crisbeto commented:
That file contains the all the retained symbols within a particular bundle (called `forms_reactive` in this case). Since the bundle includes both `core` and `platform-browser`, it shows up twice.

lpl14 commented:
Aah okay. Thanks for the explanation.

alfaproject commented:
@crisbeto since the code is basically the same (just slightly different types), maybe export the core method as a private method and use it in the other bundles for a little bit smaller bundle size?

crisbeto commented:
I was considering that, but I'm not sure that it's worth the trouble. This isn't the only place where we've duplicated smaller functions like this in order to keep the public footprint smaller.

AndrewKushnir commented:
FYI I've started an [exploratory presubmit](http://test/OCL:397217314:BASE:397217323:1631843431229:e5ac11e3) (internal-only link) to see if there are any cases that might not be covered.

AndrewKushnir commented:
FYI exploratory presubmit went well, starting a [global one](https://test.corp.google.com/OCL:397217314:BASE:397651448:1632098462209:610654eb) as well.

AndrewKushnir commented:
Quick update: there is 1 failing target in Google's codebase and more investigation is needed to see if it's an app issue or any changes in this PR are needed. No ETA on investigation results atm.

crisbeto committed:
fix(core): incorrectly inserting elements inside <template> element

Currently whenever we insert element we do it directly on the node using `appendChild` or `insertBefore`, however this doesn't work with `<template>` elements where the `template.content` has to be used.

These changes add a few checks to call `appendChild` and `insertBefore` on the `content` of the template.

Fixes #15557.

crisbeto commented:
Rebased this and re-ran the TGP. It looks like there's only one legitimate failure. I'll dig more into it.

crisbeto commented:
[Passing TGP](https://fusion2.corp.google.com/presubmit/tap/430180099/OCL:430180099:BASE:430430232:1645623325398:9054e1bf), removing the blocking label.

Merge assistance: please patch cl/430180099 which contains a fix for a target that was broken by these changes.

jessicajaniuk commented:
I've removed the merge labels until it gets fully reviewed.

alxhub commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -653,6 +655,11 @@ function nativeRemoveChild(
 
+/** Checks if an element is a `<template>` node. */
+function isTemplateNode(node: RElement): node is RTemplate {
+  return node.tagName === 'TEMPLATE' && (node as RTemplate).content !== undefined;
```
This does a megamorphic read of `.content`. Can we maybe do `node instanceof HTMLTemplateElement` instead?

crisbeto replied:
My thinking was that it would only be a megamorphic read if the element is a `<template>` *and* the browser doesn't support `template` elements. The `tagName` read is okay, because all DOM nodes have it and the `content` read will only be hit once we're relatively certain that it exists. I don't mind removing the `content` check, but I was playing it safe. The `instanceof` check would work too, but we'd also need null checking that the global class exists so that's why I didn't do it.

alxhub replied:
That makes sense - good point on the short circuiting.

jessicajaniuk commented:
Should we add `action: merge`?

crisbeto commented:
Marking as ready to merge.

Merge assistance: caretaker please patch cl/430180099 which contains a fix for a target that was broken by these changes.

AndrewKushnir commented:
This PR was merged into the repository by commit 9fa6f5a552b7780ce32aa5a60273d4008a7b6afb.
